text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,492 @@
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&'vec self) -> VecIter<'vec, T> { ... }
+}
+```
+
+However, in cases like this example, we would normally be able to leave off the
+lifetime with `&`, since there's only one source of data we could be borrowing
+from. We can do something similar with structs:
+
+```rust
+impl<T> Vec<T> {
+    fn iter(&self) -> VecIter<_, T> { ... }
+}
+```
+
+The `_` marker makes clear to the reader that *borrowing is happening*, which
+might not otherwise be clear.
+
+## `impl` blocks and lifetimes
+
+When writing an `impl` block for a structure that takes a lifetime parameter,
+you can give that parameter a name, but it must begin with an uppercase letter:
+
+```rust
+impl<T> VecIter<'Vec, T> { ... }
+```
+
+The reason for this distinction is so that there's no potential for confusion
+with lifetime names introduced by methods within the `impl` block. In other
+words, it makes clear whether the method is *introducing* a new lifetime
+(lowercase), or *referring* to the one in the `impl` header (uppercase):
+
+```rust
+impl<T> VecIter<'Vec, T> {
+    fn foo(&self) -> &'Vec T { ... }
+    fn bar(&self, arg: &'arg Bar) -> &'arg Bar { ... }
+
+    // these two are the same:
+    fn baz(&self) -> &T { ... }
+    fn baz(&'self self) -> &'self T { ... }
+}
+```
+
+If the type's lifetime is not relevant, you can leave it off using `_`:
+
+```rust
+impl<T> VecIter<_, T> { ... }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Lifetimes in `impl` headers
+
+When writing an `impl` header, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `impl<'a>`). It is also deprecated to
+use any lifetime variables that do not begin with a capital letter (which is
+needed to clearly distinguish lifetimes that are bound in the `impl` header from
+those bound in signatures).
+
+Instead the `impl` header can mention lifetimes without adding them as
+generics. **These lifetimes *always* become parameters of the `impl`**--after
+all, **no lifetime variables can possibly be in scope already**. In other words,
+the lifetime parameters of an `impl` are taken to be the set of lifetime
+variables it uses.
+
+Thus, where today you would write:
+
+```rust
+impl<'a> Iterator for MyIter<'a> { ... }
+impl<'a, 'b> SomeTrait<'a> for SomeType<'a, 'b> { ... }
+```
+
+tomorrow you would write:
+
+```rust
+impl Iterator for MyIter<'A> { ... }
+impl SomeTrait<'A> for SomeType<'A, 'B> { ... }
+```
+
+## Lifetimes in `fn` signatures
+
+When writing a `fn` declaration, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `fn foo<'a>(arg: &'a str)`).
+
+Instead:
+
+- If a lowercase lifetime variable occurs anywhere in the signature, it is
+  *always bound* by the function (as if it were in the `<>` bindings).
+- If an uppercase lifetime variable occurs, it is *always a reference* to a
```

Comment:
*always* is a strong word: I guess you want to implement it through an error that can't be turned off?

Generally I'm all for enforcing naming rules more strongly, but I think enforcement should be consistent, so IMO we should add this as warn-by-default lint to the `bad_style` lint group like the other lints that concern naming styles.